src/Type.cpp

```diff
@@ -7,33 +7,19 @@ namespace Halide {
 using std::ostringstream;
 
 /** Return an integer which is the maximum value of this type. */
-int Type::imax() const {
+int64_t Type::imax() const {
     if (is_uint()) {
-        if (bits == 32) {
-            return 0xffffffff;
-        } else if (bits < 32) {
-            return (int) ((1 << bits) - 1);
+        if (bits == 64) {
+            return -1;
         } else {
-            internal_error
-                << "Can't call Type::imax() on " << (*this)
-                << " because value is too large to represent as a signed 32-bit integer\n";
-            return 0;
+            return (1LL << bits) - 1;
         }
     } else if (is_int()) {
-        if (bits == 32) {
-            return 0x7fffffff;
-        } else if (bits < 32) {
-            return (int) ((1 << (bits-1)) - 1);
-        } else {
-            internal_error
-                << "Can't call Type::imax() on " << (*this)
-                << " because value is too large to represent as a signed 32-bit integer\n";
-            return 0;
-        }
+        return (1LL << (bits-1)) - 1;
```

Comment:
Would be better written as "(int64_t)((1UL << (bits-1)) - 1)" as "(1LL << 63) - 1" results in signed overflow?